R/Viewer.R

```diff
@@ -0,0 +1,232 @@
+        stringsAsFactors = FALSE))
+    }
+  }
+
+  # just return a list of the objects and their types, possibly filtered by the
+  # options above
+  data.frame(
+    name = objs[["table_name"]],
+    type = tolower(objs[["table_type"]]),
+    stringsAsFactors = FALSE
+  )
+}
+
+list_columns <- function(connection, ...) {
+  UseMethod("list_columns")
+}
+
+list_columns.default <- function(connection, ...) {
+  # default implementation: list columns in the given table
+  args <- list(...)
+
+  # read table or view as desired
+  obj <- if (is.null(args$table)) args$view else args$table
+
+  # specify schema or catalog if given
+  cols <- connection_sql_columns(connection$con@ptr,
+    table_name = obj,
+    schema_name = if (is.null(args$schema)) "" else args$schema,
+    catalog_name = if (is.null(args$catalog)) "" else args$catalog)
+
+  # extract and name fields for observer
+  data.frame(
+    name = cols[["name"]],
+    type = cols[["field.type"]],
+    stringsAsFactors = FALSE)
+}
+
+preview_object <- function(connection, rowLimit, ...) {
+  UseMethod("preview_object")
+}
+
+preview_object.default <- function(connection, rowLimit, ...) {
+  args <- list(...)
+
+  # read table or view as desired
+  obj <- if (is.null(args$table)) args$view else args$table
+
+  # append schema if specified
+  if (!is.null(args$schema))
+    obj <- paste(args$schema, obj, sep = ".")
+
+  dbGetQuery(connection$con, paste("SELECT * FROM", obj))
+}
+
+connection_icon <- function(connection) {
+  UseMethod("connection_icon")
+}
+
+connection_icon.default <- function(connection) {
+  # no icon is returned by default
+  ""
+}
+
+connection_actions <- function(connection) {
+  UseMethod("connection_actions")
+}
+
+connection_actions.default <- function(connection) {
+  list(
+    Help = list(
+      # show README for this package as the help; we will update to a more
+      # helpful (and/or more driver-specific) website once one exists
+      icon = "",
+      callback = function() {
+        utils::browseURL("https://github.com/rstats-db/odbc/blob/master/README.md")
+      }
+    )
+  )
+}
+
+on_connection_closed <- function(con) {
+  # make sure we have an observer
+  observer <- getOption("connectionObserver")
+  if (is.null(observer))
+    return(invisible(NULL))
+
+  type <- con@info$dbms.name
+  host <- con@info$dbname
+  observer$connectionClosed(type, host)
+}
+
+on_connection_updated <- function(con, hint) {
+  # make sure we have an observer
+  observer <- getOption("connectionObserver")
+  if (is.null(observer))
+    return(invisible(NULL))
+
+  type <- con@info$dbms.name
+  host <- con@info$dbname
+  observer$connectionUpdated(type, host, hint = hint)
+}
+
+on_connection_opened <- function(con, code) {
+  # make sure we have an observer
+  observer <- getOption("connectionObserver")
+  if (is.null(observer))
+    return(invisible(NULL))
+
+  # wrap the S4 connection object in an S3 object which understands the
```

Comment:
I may be refactoring the connection objects to be S4 subclasses of `OdbcConnection`, since we run into this same issue (wanting to dispatch on the `dbms.name`) in dplyr as well.